flfm/app/main.py

```diff
@@ -0,0 +1,470 @@
+    # and ``n_frames`` to the js transform function (and have it dynamically served/reloaded).
+    tooltip["template"] = slider_template_str(value, _center_frame(n_frames), step_size)
+    fig = plot_image(image_data[id["index"]][value, :, :], color_scale=color_scale)
+    return fig, tooltip
+
+
+@callback(
+    Output({"type": "image-slider", "index": ALL}, "tooltip", allow_duplicate=True),
+    Input("depth-step-size", "value"),
+    State({"type": "image-slider", "index": ALL}, "value"),
+    State({"type": "image-slider", "index": ALL}, "tooltip"),
+    State({"type": "n-frames", "index": "psf"}, "data"),
+    prevent_initial_call=True,
+)
+def update_slider_tooltip(depth_step_size, values, tooltips, n_frames):
+    for i, value in enumerate(values):
+        tooltips[i]["template"] = slider_template_str(value, _center_frame(n_frames), depth_step_size)
+    return tooltips
+
+
+def _display_image_with_slider(data, id):
+    """Plot image and display frame/depth slider (where appropriate)."""
+    if data is None:
+        return no_update
+
+    n_frames = len(data)
+    center_frame = _center_frame(n_frames)
+    default_slider_value = center_frame
+    fig = plot_image(data[center_frame, :, :])
+    children = [
+        dcc.Graph(id=dict(type="image-graph", index=id), figure=fig, responsive=True, style={"height": "100vh"}),
+        html.Label("Select Image Frame/Depth"),
+        dcc.Slider(
+            0,
+            len(data) - 1,
+            step=1,
+            value=default_slider_value,
+            id=dict(type="image-slider", index=id),
+            disabled=n_frames <= 1,
+            tooltip={
+                "always_visible": n_frames > 1,
+                "template": slider_template_str(default_slider_value, center_frame, app_settings.DEPTH_STEP_SIZE),
+            },
+        ),
+        html.Label("Color Scale"),
+        dcc.Dropdown(
+            id=dict(type="color-scale", index=id),
+            options=colorscales,
+            value=app_settings.IMSHOW_COLOR_SCALE,
+        ),
+    ]
+    return children, n_frames
+
+
+def upload_file(contents, filename):
+    """Read in data/image files."""
+    if contents is not None:
+        logger.info(f"Uploading {filename}...")
+        content_type, content_string = contents.split(",", maxsplit=1)
+        decoded = base64.b64decode(content_string)
+        bytestream = io.BytesIO(decoded)
+        data = flfm_io.open(bytestream)
+        logger.info(f"Uploading {filename} completed.")
+        return data
+
+
+@callback(
+    Output(dict(type="display", index="reconstruction"), "children", allow_duplicate=True),
+    Input("run-reconstruction", "n_clicks"),
+    Input("rl-iters", "value"),
+    State("center_x", "value"),
+    State("center_y", "value"),
+    State("radius", "value"),
+    prevent_initial_call=True,
+)
+def reconstruct(n_clicks, rl_iters, center_x, centre_y, radius):
```

Comment:
Can you add some type hints here if the types are consistent and reasonable to reference?